component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java

```diff
@@ -0,0 +1,628 @@
+     */
+    public static SignedJWT getSignedJWT(String subjectToken) throws IdentityOAuth2Exception {
+
+        SignedJWT signedJWT;
+        if (StringUtils.isEmpty(subjectToken)) {
+            return null;
+        }
+        try {
+            signedJWT = SignedJWT.parse(subjectToken);
+            logJWT(signedJWT);
+        } catch (ParseException e) {
+            throw new IdentityOAuth2Exception("Error while parsing the JWT", e);
+        }
+        return signedJWT;
+    }
+
+    /**
+     * Retrieve the JWTClaimsSet from the SignedJWT.
+     *
+     * @param signedJWT SignedJWT object
+     * @return JWTClaimsSet
+     * @throws IdentityOAuth2Exception Error when retrieving the JWTClaimsSet
+     */
+    public static JWTClaimsSet getClaimSet(SignedJWT signedJWT) throws IdentityOAuth2Exception {
+
+        JWTClaimsSet claimsSet = null;
+        try {
+            claimsSet = signedJWT.getJWTClaimsSet();
+            if (claimsSet == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Claim values are empty in the given JSON Web Token");
+            }
+        } catch (ParseException e) {
+            handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Error when retrieving claimsSet from the JWT");
+        }
+        return claimsSet;
+    }
+
+    /**
+     * Get the IdP configurations by issuer.
+     *
+     * @param jwtIssuer         Issuer of the JWT
+     * @param tenantDomain      Tenant Domain
+     * @return IdentityProvider
+     * @throws IdentityOAuth2Exception error when retrieving the IdP configurations
+     */
+    public static IdentityProvider getIdPByIssuer(String jwtIssuer, String tenantDomain) throws
+            IdentityOAuth2Exception {
+
+        IdentityProvider identityProvider = null;
+        try {
+            identityProvider = IdentityProviderManager.getInstance().getIdPByMetadataProperty(
+                    IdentityApplicationConstants.IDP_ISSUER_NAME, jwtIssuer, tenantDomain, false);
+            if (identityProvider == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("IDP not found when retrieving for IDP using property: " +
+                            IdentityApplicationConstants.IDP_ISSUER_NAME + " with value: " + jwtIssuer +
+                            ". Attempting to retrieve IDP using IDP Name as issuer.");
+                }
+                identityProvider = IdentityProviderManager.getInstance().getIdPByName(jwtIssuer, tenantDomain);
+            }
+        } catch (IdentityProviderManagementException e) {
+            handleException("Error while getting the Federated Identity Provider", e);
+        }
+        if (identityProvider != null) {
+            // if no IDPs were found for a given name, the IdentityProviderManager returns a dummy IDP with the
+            // name "default". We need to handle this case.
+            if (StringUtils.equalsIgnoreCase(identityProvider.getIdentityProviderName(), Constants
```

Comment:
What's the reason for using resident idp if there are no Idps by given name? feels wrong to do that

Reply:
Removed that logic and ignored checking file based idps